core/node/state_keeper/src/batch_executor/traits.rs

```diff
@@ -0,0 +1,238 @@
+    ) -> VmTransactionOutput {
+        // Executing a next transaction means that a previous transaction was either rolled back (in which case its snapshot
+        // was already removed), or that we build on top of it (in which case, it can be removed now).
+        self.pop_snapshot_no_rollback();
+        // Save pre-execution VM snapshot.
+        self.make_snapshot();
+
+        inspect_transaction(self, tx, trace_calls, false)
+            .unwrap_or_else(VmTransactionOutput::from_err)
+    }
+
+    fn inspect_transaction_with_optional_compression(
+        &mut self,
+        tx: Transaction,
+        trace_calls: TraceCalls,
+    ) -> VmTransactionOutput {
+        // Executing a next transaction means that a previous transaction was either rolled back (in which case its snapshot
+        // was already removed), or that we build on top of it (in which case, it can be removed now).
+        self.pop_snapshot_no_rollback();
+        // Save pre-execution VM snapshot.
+        self.make_snapshot();
+
+        // Note, that the space where we can put the calldata for compressing transactions
+        // is limited and the transactions do not pay for taking it.
+        // In order to not let the accounts spam the space of compressed bytecodes with bytecodes
+        // that will not be published (e.g. due to out of gas), we use the following scheme:
+        // We try to execute the transaction with compressed bytecodes.
+        // If it fails and the compressed bytecodes have not been published,
+        // it means that there is no sense in polluting the space of compressed bytecodes,
+        // and so we re-execute the transaction, but without compression.
+
+        if let Ok(output) = inspect_transaction(self, tx.clone(), trace_calls, true) {
+            return output;
+        }
+
+        // Roll back to the snapshot just before the transaction execution taken in `Self::execute_tx()`
+        // and create a snapshot at the same VM state again.
+        self.rollback_to_the_latest_snapshot();
+        self.make_snapshot();
+
+        inspect_transaction(self, tx.clone(), trace_calls, false)
+            .expect("Compression can't fail if we don't apply it")
+    }
+
+    fn rollback_last_transaction(&mut self) {
+        self.rollback_to_the_latest_snapshot();
+    }
+
+    fn start_new_l2_block(&mut self, l2_block: L2BlockEnv) {
+        VmInterface::start_new_l2_block(self, l2_block);
+    }
+
+    fn finish_batch(&mut self) -> FinishedL1Batch {
+        VmInterface::finish_batch(self)
+    }
+}
+
+/// VM factory used by the main batch executor. Encapsulates the storage type used by the executor
+/// so that it's opaque from the implementor's perspective.
+pub(super) trait BatchVmFactory<S>: fmt::Debug + Send + Sync {
+    fn create_vm<'a>(
+        &self,
+        l1_batch_params: L1BatchEnv,
+        system_env: SystemEnv,
+        storage: StoragePtr<StorageView<S>>,
+    ) -> Box<dyn BatchVm + 'a>
```

Comment:
If performance of a dynamically dispatched `BatchVm` is a concern, this could be replaced with a non-object safe trait using GATs:

```rust
trait BatchVmFactory {
    type Vm<S: ReadStorage>: BatchVm;

    fn create_vm<S: ReadStorage>(&self, ..) -> Self::Vm<S>;
}
```

Reply:
I don't think it's a concern, at least for now.

Reply:
You could add a new method to BatchVm instead of the factory if you don't want the BatchVm to be object safe.

Reply:
It's not that I don't want it to be object-safe (I do want it because it would allow to define custom impls in the node framework). But a boxed VM instance could theoretically lead to performance degradation due to dynamic dispatch, less opportunities for inlining / contextual optimizations etc.